### ISSUE TITLE
[HPOS] Unable to purchase a subscription via BACS checkout

### ISSUE THREAD

**Issue by Jinksi** [Automattic/woocommerce-subscriptions-core#265]
## Describe the bug

When purchasing a subscription via BACS checkout with HPOS enabled on the current `trunk` branch, the checkout will not be successful. The response from `/?wc-ajax=checkout` will either timeout or return a `500` status.

Activating additional plugins (e.g. Query Monitor) results in the following memory error:

```
CRITICAL Allowed memory size of 134217728 bytes exhausted (tried to allocate 1052672 bytes) in /var/www/html/wp-content/plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php on line 111
```

## To Reproduce

0. Install WC 7.1.0 and WC Subs 4.6.0
1. Checkout and activate subscriptions-core 5.0.0
3. Enable HPOS
4. Purchase a simple subscription purchase via checkout, using BACS
5. Checkout should be successful
6. Checkout subscriptions-core `trunk` branch
7. Purchase a simple subscription purchase via checkout, using BACS
8. The checkout will not be successful. The response from `/?wc-ajax=checkout` will either timeout or return a `500` status.

## Product impact

- [ ] Does this issue affect WooCommerce Subscriptions? yes/no/tbc, add issue ref
- [ ] Does this issue affect WooCommerce Payments? yes/no/tbc, add issue ref

## Additional context

I use a `wp-env` environment with WC 7.1.0 & WC Subs 4.6.0.

This error does not occur when using subs-core v5.0.0, therefore it may have been introduced by #256.

<details>

```
### WordPress Environment ###

WordPress address (URL): http://localhost
Site address (URL): http://localhost
WC Version: 7.1.0
REST API Version: ✔ 7.1.0
WC Blocks Version: ✔ 8.7.5
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.54 (Debian)
PHP Version: 7.4.33
PHP Post Max Size: 8 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.74.0
OpenSSL/1.1.1n

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.9.4-MariaDB-1:10.9.4+maria~ubu2204
Max Upload Size: 2 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.1.0
WC Database Prefix: wp_
Total Database Size: 7.32MB
Database Data Size: 5.43MB
Database Index Size: 1.89MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.09MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.08MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 4.03MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.17MB + Index: 0.19MB + Engine InnoDB
wp_posts: Data: 0.05MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.05MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
page: 7
post: 2
product: 2
shop_order: 42
shop_subscription: 8
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (4) ###

WooCommerce Subscriptions Core: by Automattic – 5.0.0
WooCommerce Subscriptions: by WooCommerce – 4.6.0
WooCommerce: by Automattic – 7.1.0
WP Console: by Edi Amin – 2.3.1

### Inactive Plugins (4) ###

Hello Dolly: by Matt Mullenweg – 1.7.2
Query Monitor: by John Blackbourn – 3.10.1
WooCommerce Payments: by Automattic – 5.0.3
WooCommerce Payments Dev Tools: by Automattic –

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 4.1.3 (update to version 4.2.0 is available)
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Subscriptions ###

WCS_DEBUG: ✔ No
Subscriptions Mode: ❌ Staging
Subscriptions Live URL: http://localhost
Subscription Statuses: wc-active: 2
wc-pending: 3
wc-on-hold: 1
wc-cancelled: 2

WooCommerce Account Connected: ❌ No
Report Cache Enabled: ✔ Yes
Cache Update Failures: ✔ 0 failure

### Store Setup ###

Country / State: United States (US) — Massachusetts

### Subscriptions by Payment Gateway ###

Direct bank transfer: wc-active: 1
wc-cancelled: 2
wc-on-hold: 1
wc-pending: 3

other: wc-active: 1

### Payment Gateway Support ###

Direct bank transfer: products

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
multichannel-marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
settings

Daily Cron: ✔ Next scheduled: 2022-11-16 03:45:05 +00:00
Options: ✔
Notes: 51
Onboarding: completed

### Action Scheduler ###

Complete: 240
Oldest: 2022-10-24 06:12:15 +0000
Newest: 2022-11-15 23:49:25 +0000

Pending: 1
Oldest: 2022-11-15 23:59:26 +0000
Newest: 2022-11-15 23:59:26 +0000


### Status report information ###

Generated at: 2022-11-15 23:54:57 +00:00
```

</details>

**Comment by brucealdridge** [comment]
It seems to be going into a recursive loop when trying to `WC_Subscription::get_date_paid()`  which gets related orders which somehow must be returning the current subscription as a related order. (Or the related order is fetching the related orders, which includes the subscription) 

Might be related to #256 ?

full investigation and logs: p1668556902424839-slack-C03SKBVPQ78

**Comment by Jinksi** [comment]
Another issue that seems to be related: a critical error when viewing the WC → Subscriptions page with HPOS enabled.

This doesn't occur when either HPOS is disabled or using the `5.0.0` branch.

<img width="1144" alt="image" src="https://user-images.githubusercontent.com/3147296/202099305-6d6f2aae-198e-44d9-bccc-059cd700dacf.png">

**Comment by haszari** [comment]
Is this issue blocking some of us from testing & making progress? (e.g. is this critical blocker for high performance order storage project)

If so, who has the ball for fixing?

**Comment by Jinksi** [comment]
> Is this issue blocking some of us from testing & making progress? (e.g. is this critical blocker for high performance order storage project)
> If so, who has the ball for fixing?

@haszari Yes, @james-allan has prepared a fix, ~and I'm creating a PR for it now~: #271.

**Update**: @james-allan is working on the PR: #272.
